provider/terraform/custom_code.rb

```diff
@@ -0,0 +1,157 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'api/object'
+require 'provider/abstract_core'
+require 'provider/property_override'
+
+module Provider
+  class Terraform < Provider::AbstractCore
+    # Inserts custom code into terraform resources.
+    class CustomCode < Api::Object
+      # Collection of fields allowed in the CustomCode section for
+      # Terraform.
+
+      # ======================
+      # schema.Resource stuff
+      # ======================
+      # Extra Schema Entries go below all other schema entries in the
+      # resource's Resource.Schema map.  They should be formatted as
+      # entries in the map, e.g. `"foo": &schema.Schema{ ... },`.
+      attr_reader :extra_schema_entry
+      # Resource definition code is inserted below everything else
+      # in the resource's Resource {...} definition.  This may be useful
+      # for things like a MigrateState / SchemaVersion pair.
+      # This is likely to be used rarely and may be removed if all its
+      # use cases are covered in other ways.
+      attr_reader :resource_definition
+      # ====================
+      # Encoders & Decoders
+      # ====================
+      # The encoders are functions which take the `obj` map after it
+      # has been assembled in either "Create" or "Update" and mutate it
+      # before it is sent to the server.  There are lots of reasons you
+      # might want to use these - any differences between local schema
+      # and remote schema will be placed here.
+      # Because the call signature of this function cannot be changed,
+      # the template will place the function header and closing } for
+      # you, and your custom code template should *not* include them.
+      attr_reader :encoder
+      # The update encoder is the encoder used in Update - if one is
+      # not provided, the regular encoder is used.  If neither is
+      # provided, of course, neither is used.  Similarly, the custom
+      # code should *not* include the function header or closing }.
+      # Update encoders are only used if object.input is false,
+      # because when object.input is true, only individual fields
+      # can be updated - in that case, use a custom expander.
+      attr_reader :update_encoder
+      # The decoder is the opposite of the encoder - it's called
+      # after the Read succeeds, rather than before Create / Update
+      # are called.  There can be many decoders - the decoder
+      # attribute is a hash from attribute name to function.  A
```

Comment:
Can you add something to explain why encoders are per object but decoders are per attribute?

Reply:
I'm having a hard time justifying it.  It turns out a lot cleaner for `Disk` because there's lots of global stuff you need for the Encoder, but lots of stuff that you can safely duplicate across attributes for `Decoder`.  Options: add per-attribute encoders and global decoders as well, or standardize on global for both.  Thoughts?

Reply:
Since it's custom code, I think it makes sense to do global for both - you can always write a clean decode function that has reusable elements in it, and I'd rather keep the total number of injection points low for now (since they're probably easier to add than to remove). It's very personal-preference-ey though; I don't have a particularly strong opinion.

Reply:
Okay!  Done.